### ISSUE TITLE
Error after pattern matching

### ISSUE THREAD

**Issue by martinescardo** [agda/agda#6719]
Consider the following two modules A and B (each in its own file).

```agda
open import Agda.Primitive

module Issue.A (i : Level) where

 data D : Set (lsuc i) where
  C : Set i → D
```

```agda
open import Agda.Primitive

module Issue.B (i : Level) where

open import Issue.A i

test : D → D
test (Issue.A.C x) = {!!} -- Doesn't type check.
```

The definition of `test` is generated by pattern matching with `ctrl-c ctrl-c`, and then it gives an error. 

We can "solve" this as follows:
```agda
open import Agda.Primitive

module Issue.B-prime (i : Level) where

import Issue.A
open Issue.A i

test : D → D
test (Issue.A.C x) = {!!} -- Now this type checks.
```
But this is not really a solution, because the file names and modules and constructors `Issue`, `A` and `C` may be rather long in practice. Moreover, if we edit this manually to
```agda
test (C x) = {!!}
```
then this type checks both in `B` and in `B-prime`.

I don't remember this problem happening before, but it does happen in 2.6.2 and 2.6.3. 

Adding @tomdjong as we were discussing the issue before submitting it here.

**Comment by jespercockx** [comment]
This is already reported in https://github.com/agda/agda/issues/3209. I agree that it is very annoying, but fixing it properly requires a overhaul of how Agda handles modules internally. See the recent master thesis by Ivar de Bruin at http://resolver.tudelft.nl/uuid:98b8fbf5-33f0-4470-88b0-39a9d526b115, in particular section 5.3.2.1.

**Comment by martinescardo** [comment]
But it also means that `open import` is not the same thing as `import` and then `open`.

**Comment by jespercockx** [comment]
Yes this is perhaps a separate issue, maybe related to https://github.com/agda/agda/issues/3240.

**Comment by martinescardo** [comment]
For the record, here is a workaround if somebody lands here and sees that there are no plans to fix this:
```agda
open import Agda.Primitive

module Issue.B-workaround (i : Level) where

open import Issue.A renaming (D to D')

D = D' i

test : D → D
test (C x) = {!!} -- Produced right pattern with ctrl-c ctrl-c.
```

**Comment by martinescardo** [comment]
In that case, it is better to make the parameter `i` implicit, so that all the other functions in `A` don't need to be given `i` explicitly as a parameter.

**Comment by martinescardo** [comment]
OK. I found a way out that is more transparent and less hacky.

Create a new file with the definition of the data type in A with all parameters given explicitly.

Then in the module `A` above call that module to define D. Everything works well after that.

Moral: don't define a `data` type in a module with parameters if you want your pattern matching to succeed and work well. Only define `data` types that don't depend on module parameters.

I adopted this in `TypeTopology/source/Iterative`, which sorts out the headaches I had in the last few days. 

But I wish to record this here, not only for others but also for myself, if I forget the workaround in the future.
